src/arithmetic.rs

```diff
@@ -180,6 +180,39 @@ fn multiexp_serial<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C], acc: &mut
     }
 }
 
+/// Performs a small multi-exponentiation operation.
+/// Uses the double-and-add algorithm with doublings shared across points.
+
+pub fn small_multiexp<C: CurveAffine>(coeffs: &[C::Scalar], bases: &[C]) -> C::Projective {
+    // Gets the bit at position `i`. Bits are numbered from 0 (least significant) to 7 (most significant).
+    fn get_bit_at(byte: u8, i: usize) -> bool {
```

Comment:
I think it'd be better to just inline this function as `((byte >> bit_idx) & 1) == 1` at the callsite rather than introducing a new function for it.